### PR TITLE
feat/reputation

### PR DESCRIPTION
**Pull request by acharl**

### CODE REVIEW

File: pallets/marketplace/src/lib.rs [hunk 1]
```diff
@@ -357,6 +392,54 @@ pub mod pallet {
...
+                .try_get_amount()
+                .map_err(|_| Error::<T>::JobRegistrationUnsupportedReward)?
+                .into();
+
+            let avg_job_reward = <StoredAvgJobReward<T>>::get(&reward_asset).unwrap_or_default();
+            let total_jobs_assigned =
+                <StoredTotalJobsAssigned<T>>::get(&reward_asset).unwrap_or_default();
+
+            let total_rewards = avg_job_reward
+                .checked_mul(&(total_jobs_assigned - <T as Config>::AssetAmount::from(1))) // NEXT checked_sub?
```
Comment (gitsimon) [inline pallets/marketplace/src/lib.rs:413]: see above: this should be an unsigned integer not an AssetAmount

File: pallets/marketplace/src/lib.rs [hunk 2]
```diff
@@ -357,6 +392,54 @@ pub mod pallet {
...
+                .into();
+
+            let reward_amount: T::AssetAmount = extra
+                .reward
+                .try_get_amount()
+                .map_err(|_| Error::<T>::JobRegistrationUnsupportedReward)?
+                .into();
+
+            let avg_job_reward = <StoredAvgJobReward<T>>::get(&reward_asset).unwrap_or_default();
+            let total_jobs_assigned =
```
Comment (gitsimon) [inline pallets/marketplace/src/lib.rs:409]: Should it be default if not yet in map? Or rather ensure it is 0 by using 0.into()

File: pallets/marketplace/src/lib.rs [hunk 3]
```diff
@@ -79,11 +91,28 @@ pub mod pallet {
...
     /// The storage for remaining capacity for each source. Can be negative if capacity is reduced beyond the number of jobs currently assigned.
     #[pallet::storage]
     #[pallet::getter(fn stored_capacity)]
     pub type StoredCapacity<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, i32>;
 
+    /// Number of total jobs assigned as a map [AssetId] -> AssetAmount>
+    #[pallet::storage]
+    #[pallet::getter(fn total_jobs_assigned)]
+    pub type StoredTotalJobsAssigned<T: Config> =
+        StorageMap<_, Blake2_128Concat, <T as Config>::AssetId, <T as Config>::AssetAmount>;
```
Comment (gitsimon) [inline pallets/marketplace/src/lib.rs:109]: the key should be an unsigned integer not an AssetAmount

File: pallets/marketplace/src/lib.rs [hunk 4]
```diff
@@ -466,6 +563,14 @@ pub mod pallet {
                             JobStatus::Assigned,
                         );
                     }
+                    let total_jobs_assigned =
+                        <StoredTotalJobsAssigned<T>>::get(&reward_asset).unwrap_or_default();
+                    let new_total_jobs_assigned = total_jobs_assigned + T::AssetAmount::from(1);
```
Comment (gitsimon) [inline pallets/marketplace/src/lib.rs:568]: see above: this should be an unsigned integer not an AssetAmount

File: pallets/marketplace/src/lib.rs [hunk 5]
```diff
@@ -466,6 +563,14 @@ pub mod pallet {
...
+
+                    <StoredTotalJobsAssigned<T>>::insert(
+                        reward_asset.clone(),
+                        new_total_jobs_assigned,
+                    );
```
Comment (gitsimon) [inline pallets/marketplace/src/lib.rs:573]: Nit-picky: you could have it in one operation like

<StoredTotalJobsAssigned<T>>::mutate(&reward_asset, |total_jobs_assigned| {
  *total_jobs_assigned = total_jobs_assigned + 1u128
});

**Review by gitsimon — CHANGES_REQUESTED** [review]

File: pallets/marketplace/Cargo.toml [hunk 6]
```diff
@@ -38,6 +38,12 @@ hex-literal = { version = "0.3", optional = true }
 parachain-info = { git = "https://github.com/paritytech/cumulus", default-features = false, optional = true, branch = "polkadot-v0.9.29" }
 pallet-balances = { git = "https://github.com/paritytech/substrate", default-features = false, optional = true, branch = "polkadot-v0.9.29" }
 sp-core = { git = "https://github.com/paritytech/substrate", default-features = false, optional = true, branch = "polkadot-v0.9.29" }
+
+# reputation
+reputation = { path = "../../reputation" }
```
[the remark this answers is not in the document]
Comment (acharl) [reply pallets/marketplace/Cargo.toml]: agreed, done in d3a72618ca2b37fa1c2b743aff2fb95bcaf05710

File: pallets/marketplace/src/lib.rs [hunk 7]
```diff
@@ -79,11 +91,28 @@ pub mod pallet {
...
     #[pallet::getter(fn stored_capacity)]
     pub type StoredCapacity<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, i32>;
 
+    /// Number of total jobs assigned as a map [AssetId] -> AssetAmount>
+    #[pallet::storage]
+    #[pallet::getter(fn total_jobs_assigned)]
+    pub type StoredTotalJobsAssigned<T: Config> =
+        StorageMap<_, Blake2_128Concat, <T as Config>::AssetId, <T as Config>::AssetAmount>;
+
+    /// Average job reward as a map [AssetId] -> AssetAmount>
```
[the remark this answers is not in the document]
Comment (acharl) [reply pallets/marketplace/src/lib.rs]: done in 423b6de31b2784530765cef665d5a149e8d00d92

File: pallets/marketplace/src/lib.rs (hunk 4 above)
Comment (acharl) [reply pallets/marketplace/src/lib.rs]: agreed, fixed in https://github.com/Acurast/acurast-core/commit/3dfc18721e98aa5f02200efdefe7d42bbb67e5da

File: pallets/marketplace/src/lib.rs (hunk 2 above)
Comment (acharl) [reply pallets/marketplace/src/lib.rs:408]: done in 9163ba3fcc510ebb04b4ec52d5456b9566943a19

File: pallets/marketplace/src/lib.rs [hunk 8]
```diff
@@ -47,20 +48,31 @@ pub mod pallet {
...
         type AssetAmount: Parameter
             + CheckedMul
+            + CheckedDiv
+            + CheckedAdd
+            + CheckedSub
             + From<u128>
+            + Into<u128>
+            + Default
             + Ord
+            + Copy
```
[the remark this answers is not in the document]
Comment (acharl) [reply pallets/marketplace/src/lib.rs]: agreed, addressed in 3d5d0f223a0b27d9cf16510c592b110b12eed709

**Review by godenzim — APPROVED** [review]

**Pull request merged**